### PR TITLE
Boussinesq approximation to incompressible navier stokes

### PR DESCRIPTION
**Pull request by gridley**
This PR implements a kernel that provides the residual:

<a href="https://www.codecogs.com/eqnedit.php?latex=g&space;\alpha&space;\Delta&space;T&space;\psi" target="_blank"></a>

Where psi is the test function, g is the acceleration due to gravity, and alpha is the linearized expansion coefficient about some nominal temperature given by:

<a href="https://www.codecogs.com/eqnedit.php?latex=\rho&space;=&space;\rho_0&space;(1-\alpha&space;\Delta&space;T)" target="_blank"></a>

To compute delta T, an aux kernel called ConstantDifferenceAux was added, which sets an aux variable by subtracting some value from a problem variable.

Two examples were added in tests/laminar_flow. Boussinesq cube has been run on Blue Waters and looks reasonable. Boussinesq square can easily be run on a desktop computer. A video of boussinesq square can be found at:

https://youtu.be/wNoeH50IClM

A future improvement could be fixing up the Boussinesq Kernel off-diagonal jacobian to support variable material properties.

### CODE REVIEW

File: src/auxkernels/ConstantDifferenceAux.C [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  return params;
+}
+
+/**
+ * The original intention of this aux kernel is to compute a delta T about some
+ * nominal temperature used in the Boussinesq approximation for buoyancy in
+ * incompressible Navier Stokes. The nominal temperature may need to change over
+ * time in a simulation so that the relative size of the residual contributed by
+ * the INSBoussinesqBodyForce stays around the same size as the other kernels.
+ */
```
Comment (lindsayad) [inline src/auxkernels/ConstantDifferenceAux.C]: This doxygen comment should go into the header file before the class declaration (not the forward declaration)

File: src/kernels/INSBoussinesqBodyForce.C [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  return 0;
+}
+
+Real
+INSBoussinesqBodyForce::computeQpOffDiagJacobian(unsigned jvar)
+{
+  if (jvar == _dT_id)
+    // not a perfect one, this could include d alpha dT and d rho dT
+    // but those oughta be small
+    return -_test[_i][_qp] * _alpha[_qp] * _gravity(_component) * _rho[_qp];
```
Comment (lindsayad) [inline src/kernels/INSBoussinesqBodyForce.C]: Adding this really gave you better convergence? dT = T - constant, from your AuxKernel, right? In which case, I would expect the jacobian to look like this:
```c++
return _test[_i][_qp] * _alpha[_qp] * _gravity(_component) * _rho[_qp] * _phi[_j][_qp];
```
e.g. the sign of the residual and the jacobian should be the same. Additionally you are missing `_phi`

File: src/auxkernels/ConstantDifferenceAux.C (hunk 1 above)
Comment (gridley) [inline src/auxkernels/ConstantDifferenceAux.C]: done, thanks.

File: src/kernels/INSBoussinesqBodyForce.C (hunk 2 above)
Comment (gridley) [inline src/kernels/INSBoussinesqBodyForce.C]: Ah, right, didn't realize that you differentiate with respect to the nodal value. Makes sense.

**Comment by gridley** [comment]
A test with 5 timesteps now happens:

`laminar_flow.boussinesqSquare............................................................................. OK
-------------------------------------------------------------------------------------------------------------
Ran 6 tests in 25.4 seconds
6 passed, 1 skipped, 0 pending, 0 failed
`

**Pull request merged**